From af1e622be02e7167a8a19082f5533039ed8dce1d Mon Sep 17 00:00:00 2001
From: Vitaly Miniakhmetov <vitalii.miniakhmetov@t-platforms.ru>
Date: Thu, 17 Jun 2021 12:28:46 +0500
Subject: [PATCH 055/106] i2c-baikal-smbus: Support fwnode in probe

---
 drivers/i2c/busses/i2c-baikal-smbus.c | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/drivers/i2c/busses/i2c-baikal-smbus.c b/drivers/i2c/busses/i2c-baikal-smbus.c
index a1ff5ed9cb12..045a65324a33 100644
--- a/drivers/i2c/busses/i2c-baikal-smbus.c
+++ b/drivers/i2c/busses/i2c-baikal-smbus.c
@@ -425,7 +425,11 @@ static int be_smbus_probe(struct platform_device *pdev)
 	besmb->adapter.owner = THIS_MODULE;
 	besmb->adapter.algo = &be_smbus_algorithm;
 	besmb->adapter.dev.parent = &pdev->dev;
-	besmb->adapter.dev.of_node = pdev->dev.of_node;
+	if (dev_of_node(&pdev->dev)) {
+		besmb->adapter.dev.of_node = pdev->dev.of_node;
+	} else {
+		besmb->adapter.dev.fwnode = pdev->dev.fwnode;
+	}
 
 	platform_set_drvdata(pdev, besmb);
 
-- 
2.33.0

