From 37b022337542ed94dda4cb9a8c7f13d397f665fc Mon Sep 17 00:00:00 2001
From: Vitaly Miniakhmetov <vitalii.miniakhmetov@t-platforms.ru>
Date: Thu, 16 Sep 2021 18:43:58 +0500
Subject: [PATCH 099/106] Fixed license for module mv88x2222

---
 drivers/net/phy/mv88x2222.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/net/phy/mv88x2222.c b/drivers/net/phy/mv88x2222.c
index 1d2cdd9d2d55..bf964f235e62 100644
--- a/drivers/net/phy/mv88x2222.c
+++ b/drivers/net/phy/mv88x2222.c
@@ -32,7 +32,7 @@
 #include <linux/of_gpio.h>
 
 MODULE_DESCRIPTION("Marvell Ethernet Transceiver driver");
-MODULE_LICENSE("Proprietary");
+MODULE_LICENSE("GPL");
 
 /* 31.F002 Line side mode (ch.3.1.2, pg.46) */
 #define MV_MODE_LINE_SHF			8
-- 
2.33.0

