baikal-vdu: Various fixes

From: Vitaly Miniakhmetov <vitalii.miniakhmetov@t-platforms.ru>

diff --git a/drivers/gpu/drm/baikal/baikal_vdu_crtc.c b/drivers/gpu/drm/baikal/baikal_vdu_crtc.c
index d97e6c7d944a..322b7ecf83ae 100644
--- a/drivers/gpu/drm/baikal/baikal_vdu_crtc.c
+++ b/drivers/gpu/drm/baikal/baikal_vdu_crtc.c
@@ -140,8 +140,7 @@ static void baikal_vdu_crtc_helper_mode_set_nofb(struct drm_crtc *crtc)
 
 	rate = mode->clock * 1000;
 
-	if (rate != clk_get_rate(priv->clk))
-	{
+	if (rate != clk_get_rate(priv->clk)) {
 		DRM_DEV_DEBUG_DRIVER(dev->dev, "Requested pixel clock is %lu Hz\n", rate);
 
 		/* hold clock domain reset; disable clocking */
@@ -167,7 +166,7 @@ static void baikal_vdu_crtc_helper_mode_set_nofb(struct drm_crtc *crtc)
 		DRM_ERROR("Cannot set desired pixel clock (%lu Hz)\n", rate);
 
 	ppl = mode->hdisplay / 16;
-	if (priv->panel) {
+	if (priv->panel && priv->ep_count == 2) {
 		hsw = mode->hsync_end - mode->hsync_start;
 		hfp = mode->hsync_start - mode->hdisplay - 1;
 	} else {
@@ -241,16 +240,16 @@ static void baikal_vdu_crtc_helper_enable(struct drm_crtc *crtc,
 	/* Set 16-word input FIFO watermark */
 	/* Enable and Power Up */
 	cntl = readl(priv->regs + CR1);
-	cntl &= ~CR1_FDW_MASK;
-	cntl |= CR1_LCE + CR1_FDW_16_WORDS;
+	cntl &= ~(CR1_FDW_MASK | CR1_OPS_MASK);
+	cntl |= CR1_LCE | CR1_FDW_16_WORDS;
 
 	if (priv->type == VDU_TYPE_LVDS) {
 		panel_node = panel->dev->of_node;
 		if (of_property_read_string(panel_node, "data-mapping", &data_mapping)) {
 			cntl |= CR1_OPS_LCD18;
-		} else if (strncmp(data_mapping, "vesa-24", 7))
+		} else if (!strncmp(data_mapping, "vesa-24", 7))
 			cntl |= CR1_OPS_LCD24;
-		else if (strncmp(data_mapping, "jeida-18", 8))
+		else if (!strncmp(data_mapping, "jeida-18", 8))
 			cntl |= CR1_OPS_LCD18;
 		else {
 			dev_warn(crtc->dev->dev, "%s data mapping is not supported, vesa-24 is set\n", data_mapping);
diff --git a/drivers/gpu/drm/baikal/baikal_vdu_drv.c b/drivers/gpu/drm/baikal/baikal_vdu_drv.c
index ddbb8df53b12..bd7bf1ec8024 100644
--- a/drivers/gpu/drm/baikal/baikal_vdu_drv.c
+++ b/drivers/gpu/drm/baikal/baikal_vdu_drv.c
@@ -117,21 +117,11 @@ int baikal_vdu_find_panel_or_bridge(struct device *dev,
 static int baikal_vdu_remove_efifb(struct drm_device *dev)
 {
 	int err;
-	struct apertures_struct *a;
-	a = alloc_apertures(1);
-	if (!a) {
-		err = -ENOMEM;
-		dev_warn(dev->dev, "failed to allocate apertures\n");
-		goto out;
-	}
-	a->ranges[0].base = 0;
-	a->ranges[0].size = ~0;
-	err = drm_fb_helper_remove_conflicting_framebuffers(a, "baikal-vdudrmfb", false);
-	if (err) {
+	err = drm_fb_helper_remove_conflicting_framebuffers(NULL,
+							    "baikal-vdudrmfb",
+							    false);
+	if (err)
 		dev_warn(dev->dev, "failed to remove firmware framebuffer\n");
-	}
-	kfree(a);
-out:
 	return err;
 }
 
diff --git a/drivers/gpu/drm/baikal/baikal_vdu_regs.h b/drivers/gpu/drm/baikal/baikal_vdu_regs.h
index 5553fcac5fec..dc9d7dae2f59 100644
--- a/drivers/gpu/drm/baikal/baikal_vdu_regs.h
+++ b/drivers/gpu/drm/baikal/baikal_vdu_regs.h
@@ -51,6 +51,7 @@
 #define CR1_FDW_4_WORDS     (0 << 16)
 #define CR1_FDW_8_WORDS     (1 << 16)
 #define CR1_FDW_16_WORDS    (2 << 16)
+#define CR1_OPS_MASK        GENMASK(14, 12)
 #define CR1_OPS_LCD18       (0 << 13)
 #define CR1_OPS_LCD24       (1 << 13)
 #define CR1_OPS_565         (0 << 12)
