baikal-vdu: Porting to 5.13.y

From: Vitaly Miniakhmetov <vitalii.miniakhmetov@t-platforms.ru>

diff --git a/drivers/gpu/drm/baikal/baikal_vdu_plane.c b/drivers/gpu/drm/baikal/baikal_vdu_plane.c
index 26617386855b..109ef79ba12c 100644
--- a/drivers/gpu/drm/baikal/baikal_vdu_plane.c
+++ b/drivers/gpu/drm/baikal/baikal_vdu_plane.c
@@ -23,6 +23,7 @@
 #include <drm/drm_gem_cma_helper.h>
 #include <drm/drm_gem_framebuffer_helper.h>
 #include <drm/drm_plane_helper.h>
+#include <drm/drm_gem_atomic_helper.h>
 
 #include "baikal_vdu_drm.h"
 #include "baikal_vdu_regs.h"
@@ -64,7 +65,7 @@ void baikal_vdu_update_work(struct work_struct *work)
 }
 
 static void baikal_vdu_primary_plane_atomic_update(struct drm_plane *plane,
-					      struct drm_plane_state *old_state)
+					      struct drm_atomic_state *old_state)
 {
 	struct drm_device *dev = plane->dev;
 	struct baikal_vdu_private *priv = dev->dev_private;
@@ -145,7 +146,7 @@ static void baikal_vdu_primary_plane_atomic_update(struct drm_plane *plane,
 
 static const struct drm_plane_helper_funcs baikal_vdu_primary_plane_helper_funcs = {
 	.atomic_update = baikal_vdu_primary_plane_atomic_update,
-	.prepare_fb = drm_gem_fb_prepare_fb,
+	.prepare_fb = drm_gem_plane_helper_prepare_fb,
 };
 
 static const struct drm_plane_funcs baikal_vdu_primary_plane_funcs = {
