dwc-i2s: Added deferred probe when no irq_num

From: Vitaly Miniakhmetov <vitalii.miniakhmetov@t-platforms.ru>

diff --git a/sound/soc/dwc/dwc-i2s.c b/sound/soc/dwc/dwc-i2s.c
index 38e0eaacb40e..f5b543deed79 100644
--- a/sound/soc/dwc/dwc-i2s.c
+++ b/sound/soc/dwc/dwc-i2s.c
@@ -644,7 +644,9 @@ static int dw_i2s_probe(struct platform_device *pdev)
 	dev->dev = &pdev->dev;
 	irq_num = platform_irq_count(pdev);
 
-	if (!irq_num)
+	if (irq_num < 0) /* -EPROBE_DEFER */
+		return irq_num;
+	else if (!irq_num)
 		dev_err(&pdev->dev, "No irq found on device\n");
 
 	do {
