dwc-i2s: Removed DAI probe

From: Vitaly Miniakhmetov <vitalii.miniakhmetov@t-platforms.ru>

diff --git a/sound/soc/dwc/dwc-i2s.c b/sound/soc/dwc/dwc-i2s.c
index 85b3466a1555..38e0eaacb40e 100644
--- a/sound/soc/dwc/dwc-i2s.c
+++ b/sound/soc/dwc/dwc-i2s.c
@@ -617,33 +617,6 @@ static int dw_configure_dai_by_dt(struct dw_i2s_dev *dev,
 
 }
 
-static int dw_i2s_dai_probe(struct snd_soc_dai *dai)
-{
-        struct dw_i2s_dev *dev = snd_soc_dai_get_drvdata(dai);
-        struct snd_dmaengine_dai_dma_data *playback_dma_data;
-        struct snd_dmaengine_dai_dma_data *capture_dma_data;
-
-        playback_dma_data = devm_kzalloc(dai->dev,
-                                         sizeof(*playback_dma_data),
-                                         GFP_KERNEL);
-        if (!playback_dma_data)
-                return -ENOMEM;
-
-        capture_dma_data = devm_kzalloc(dai->dev,
-                                        sizeof(*capture_dma_data),
-                                        GFP_KERNEL);
-        if (!capture_dma_data)
-                return -ENOMEM;
-
-        playback_dma_data->addr = dev->play_dma_data.pd.addr;
-        capture_dma_data->addr = dev->capture_dma_data.pd.addr;
-
-        playback_dma_data->maxburst = dev->play_dma_data.pd.max_burst;
-        capture_dma_data->maxburst = dev->capture_dma_data.pd.max_burst;
-
-        return 0;
-}
-
 static int dw_i2s_probe(struct platform_device *pdev)
 {
 	const struct i2s_platform_data *pdata = pdev->dev.platform_data;
@@ -662,7 +635,6 @@ static int dw_i2s_probe(struct platform_device *pdev)
 		return -ENOMEM;
 
 	dw_i2s_dai->ops = &dw_i2s_dai_ops;
-	dw_i2s_dai->probe = &dw_i2s_dai_probe;
 
 	res = platform_get_resource(pdev, IORESOURCE_MEM, 0);
 	dev->i2s_base = devm_ioremap_resource(&pdev->dev, res);
