panfrost: disable devfreq on BE-M1000 SoC

From: Vitaly Miniakhmetov <vitalii.miniakhmetov@t-platforms.ru>

diff --git a/drivers/gpu/drm/panfrost/panfrost_devfreq.c b/drivers/gpu/drm/panfrost/panfrost_devfreq.c
index 3644652f726f..b5321d467a7d 100644
--- a/drivers/gpu/drm/panfrost/panfrost_devfreq.c
+++ b/drivers/gpu/drm/panfrost/panfrost_devfreq.c
@@ -92,6 +92,11 @@ int panfrost_devfreq_init(struct panfrost_device *pfdev)
 	struct thermal_cooling_device *cooling;
 	struct panfrost_devfreq *pfdevfreq = &pfdev->pfdevfreq;
 
+	if (of_device_is_compatible(of_root, "baikal,baikal-m")) {
+		DRM_DEV_INFO(dev, "disabling GPU devfreq on BE-M1000\n");
+		return 0;
+	}
+
 	if (pfdev->comp->num_supplies > 1) {
 		/*
 		 * GPUs with more than 1 supply require platform-specific handling:
