Rename mitx/mitx-d to its proper names tf307-mb-s-c/d

From: Miniakhmetov Vitaly <vitalii.miniakhmetov@t-platforms.ru>

diff --git a/arch/arm64/boot/dts/baikal/Makefile b/arch/arm64/boot/dts/baikal/Makefile
index 9729c6e3d4f3..8e39db726906 100644
--- a/arch/arm64/boot/dts/baikal/Makefile
+++ b/arch/arm64/boot/dts/baikal/Makefile
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 dtb-$(CONFIG_ARCH_BAIKAL) += bm-dbm.dtb
-dtb-$(CONFIG_ARCH_BAIKAL) += bm-mbm10.dtb
-dtb-$(CONFIG_ARCH_BAIKAL) += bm-mbm20.dtb
+dtb-$(CONFIG_ARCH_BAIKAL) += edelweiss-tf307-mb-s-c.dtb
+dtb-$(CONFIG_ARCH_BAIKAL) += edelweiss-tf307-mb-s-d.dtb
 dtb-$(CONFIG_ARCH_BAIKAL) += bm-qemu.dtb
 
 always		:= $(dtb-y)
diff --git a/arch/arm64/boot/dts/baikal/bm-mbm10.dts b/arch/arm64/boot/dts/baikal/edelweiss-tf307-mb-s-c.dts
similarity index 100%
rename from arch/arm64/boot/dts/baikal/bm-mbm10.dts
rename to arch/arm64/boot/dts/baikal/edelweiss-tf307-mb-s-c.dts
diff --git a/arch/arm64/boot/dts/baikal/bm-mbm20.dts b/arch/arm64/boot/dts/baikal/edelweiss-tf307-mb-s-d.dts
similarity index 100%
rename from arch/arm64/boot/dts/baikal/bm-mbm20.dts
rename to arch/arm64/boot/dts/baikal/edelweiss-tf307-mb-s-d.dts
