samples: add Build BPF sample code

From: Roman Stavtsev <roman.stavtsev@baikalelectronics.ru>

diff --git a/samples/Kconfig b/samples/Kconfig
index b5a1a7aa7e23..c15bd7aadce8 100644
--- a/samples/Kconfig
+++ b/samples/Kconfig
@@ -223,4 +223,11 @@ config SAMPLE_WATCH_QUEUE
 	  Build example userspace program to use the new mount_notify(),
 	  sb_notify() syscalls and the KEYCTL_WATCH_KEY keyctl() function.
 
+config SAMPLE_BPF
+	bool "Build BPF sample code"
+	depends on HEADERS_INSTALL
+	help
+	  Build samples of BPF filters using various methods of
+	  BPF filter construction.
+
 endif # SAMPLES
diff --git a/samples/Makefile b/samples/Makefile
index 087e0988ccc5..b6f0419c566f 100644
--- a/samples/Makefile
+++ b/samples/Makefile
@@ -30,3 +30,4 @@ obj-$(CONFIG_SAMPLE_INTEL_MEI)		+= mei/
 subdir-$(CONFIG_SAMPLE_WATCHDOG)	+= watchdog
 subdir-$(CONFIG_SAMPLE_WATCH_QUEUE)	+= watch_queue
 obj-$(CONFIG_DEBUG_KMEMLEAK_TEST)	+= kmemleak/
+subdir-$(CONFIG_SAMPLE_BPF)             += bpf
